<a href="https://colab.research.google.com/github/andreac941/ComputerVision-Eq3/blob/main/Knapsack.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Knapsack Problem**

Es un tipo de problema de optimización combinatoria que implica seleccionar un subconjunto de elementos, cada uno con un peso y un valor específicos, para maximizar el valor total sin exceder una capacidad de peso dada.

El problema se formula comúnmente de la siguiente manera:

Dado un conjunto de elementos, cada uno con un peso wi y un valor
vi, y una mochila con una capacidad máxima W, el objetivo es encontrar la combinación más valiosa de elementos para incluir en la mochila sin exceder su capacidad de peso.

El problema de decisión se puede expresar de la siguiente manera: ¿Existe un subconjunto de elementos cuyo peso total sea menor o igual a W y cuyo valor total esté maximizado?

Ejemplo rapido de aplicacion del algoritmo:

Datos del problema:

Bloque A: Dimensiones = 3m X 2m X 1m, Precio = $100.

Bloque B: Dimensiones = 2m X 2m X 2m, Precio = $150.

Bloque C: Dimensiones = 1m X 1m X 3m, Precio = $80.

Capacidad máxima del volumen de la mochila: 15m3

Formulación del problema de la mochila:

v1 = 100, dim1 = (3,2,1)

v2 = 150, dim1 = (2,2,2)

v3 = 80, dim1 = (1,1,3)

Vmax = 5m3 (capacidad máxima del volumen de la mochila)



In [ ]:
def Mochila_0_1(dimensions, valores, capacidad_volumen):
    n = len(valores)
    matriz = [[0] * (capacidad_volumen + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        for v in range(1, capacidad_volumen + 1):
            if sum(dimensions[i-1]) <= v:
                matriz[i][v] = max(valores[i-1] + matriz[i-1][v-sum(dimensions[i-1])], matriz[i-1][v])
            else:
                matriz[i][v] = matriz[i-1][v]

    return matriz[n][capacidad_volumen]

# Datos del problema
dimensions = [(3, 2, 1), (2, 2, 2), (1, 1, 3)]
valores = [100, 150, 80]
capacidad_volumen = 15

# Aplicar la función
resultado = Mochila_0_1(dimensions, valores, capacidad_volumen)

# Imprimir el resultado
print(resultado)




250


El valor de 250 significa que, utilizando el algoritmo de la mochila adaptado para dimensiones y volumen, se puede maximizar el valor total seleccionando bloques de materia prima de manera óptima.

Si se elige cortar y seleccionar bloques de manera óptima, siguiendo las restricciones de volumen y dimensiones, se podria obtener productos cuyo valor total es de $250.

Ademas del algoritmo de Nutsack, otras formas de encontrar la solucion de un problema de optimizacion son:
1. Programacion Lineal
2. Algoritmos Geneticos
3. Algoritmos de Enjambre
4. Aprendizaje por Refuerzo
5. Redes Neuronales